File: src/__tests__/ApexListenerTest.ts
import { ApexLexer } from "../ApexLexer";
import { ApexParserListener } from "../ApexParserListener";
import { ApexParser, MethodDeclarationContext } from "../ApexParser";
import { CaseInsensitiveInputStream } from "../CaseInsensitiveInputStream"
import { CommonTokenStream } from 'antlr4ts';
import { ParseTreeWalker } from 'antlr4ts/tree/ParseTreeWalker';
import { ThrowingErrorListener } from "../ThrowingErrorListener";

class TestListener implements ApexParserListener {
    public methodCount = 0

    enterMethodDeclaration(ctx: MethodDeclarationContext): void {
        this.methodCount += 1;
    }
}

test('Listener Generates Events', () => {

    let lexer = new ApexLexer(new CaseInsensitiveInputStream("test.cls", "public class Hello {}"))
    let tokens = new CommonTokenStream(lexer);

    let parser = new ApexParser(tokens)

    parser.removeErrorListeners()
    parser.addErrorListener(new ThrowingErrorListener());

    const cu = parser.compilationUnit()
    const listener = new TestListener()
    ParseTreeWalker.DEFAULT.walk(listener as ApexParserListener, cu)

    expect(listener.methodCount == 1)
})

File: jvm/src/test/java/com/nawforce/apexparser/ApexListenerTest.java
package com.nawforce.apexparser;

import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.tree.ParseTreeWalker;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.io.StringReader;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class ApexListenerTest {

    static class TestListener extends ApexParserBaseListener {
        public Integer methodCount = 0;

        @Override
        public void enterMethodDeclaration(ApexParser.MethodDeclarationContext ctx) {
            this.methodCount += 1;
        }
    }


    @Test
    void testListenerGeneratesEvents() throws IOException {
        ApexLexer lexer = new ApexLexer(new CaseInsensitiveInputStream(new StringReader(
                "public class Hello { public void func(){} }")));
        CommonTokenStream tokens = new CommonTokenStream(lexer);
        ApexParser parser = new ApexParser(tokens);
        ApexParser.CompilationUnitContext context = parser.compilationUnit();

        TestListener listener = new TestListener();
        ParseTreeWalker.DEFAULT.walk(listener, context);

        assertEquals(listener.methodCount.intValue(), 1);
    }

}
